src/diameter_hss_connection.cpp

```diff
@@ -0,0 +1,504 @@
+/**
+ * @file diameter_hss_connection.cpp Implementation of HssConnection that uses Diameter
+ *
+ * Copyright (C) Metaswitch Networks 2017
+ * If license terms are provided to you in a COPYING file in the root directory
+ * of the source code repository by which you are accessing this code, then
+ * the license outlined in that COPYING file applies to your use.
+ * Otherwise no rights are granted except for those provided to you by
+ * Metaswitch Networks in a separate written agreement.
+ */
+
+#include "charging_addresses.h"
+#include "diameter_hss_connection.h"
+#include "homesteadsasevent.h"
+#include "servercapabilities.h"
+
+namespace HssConnection {
+
+static SNMP::CxCounterTable* mar_results_tbl;
+static SNMP::CxCounterTable* sar_results_tbl;
+static SNMP::CxCounterTable* uar_results_tbl;
+static SNMP::CxCounterTable* lir_results_tbl;
+
+template <class AnswerType>
+void DiameterHssConnection::DiameterTransaction<AnswerType>::on_response(Diameter::Message& rsp)
+{
+  update_latency_stats();
+  AnswerType answer = create_answer(rsp);
+  _response_clbk(answer);
+}
+
+template <class T>
+void DiameterHssConnection::DiameterTransaction<T>::increment_results(int32_t result,
+                                                                      int32_t experimental,
+                                                                      uint32_t vendor)
+{
+  // IMS mandates that exactly one of result code or experimental result code
+  // will be set, so we can unambiguously assume that, if one is set, then the
+  // other one won't be.
+  if (result != 0)
+  {
+    _cx_results_tbl->increment(SNMP::DiameterAppId::BASE, result);
+  }
+  else if (experimental != 0 && vendor == VENDOR_ID_3GPP)
+  {
+    _cx_results_tbl->increment(SNMP::DiameterAppId::_3GPP, experimental);
+  }
+}
+
+template <class AnswerType>
+void DiameterHssConnection::DiameterTransaction<AnswerType>::on_timeout()
+{
+  TRC_ERROR("Diameter timeout");
+
+  update_latency_stats();
+
+  // No result-code returned on timeout, so use 0.
+  _cx_results_tbl->increment(SNMP::DiameterAppId::TIMEOUT, 0);
+
+  // Call the callback with SERVER_UNAVAILABLE
+  AnswerType answer = AnswerType(ResultCode::SERVER_UNAVAILABLE);
+  _response_clbk(answer);
+}
+
+template <class T>
+void DiameterHssConnection::DiameterTransaction<T>::update_latency_stats()
+{
+  if (_stats_manager != NULL)
+  {
+    unsigned long latency = 0;
+    if (get_duration(latency))
+    {
+      if (_stat_updates & STAT_HSS_LATENCY)
+      {
+        _stats_manager->update_H_hss_latency_us(latency);
```

Comment:
I don't know, it was called this already. I think the "us" means "Homestead's view"